### PR TITLE
gh-107704: Argument Clinic: add support for deprecating keyword use of parameters

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
It is now possible to deprecate passing keyword arguments for keyword-or-positional parameters with Argument Clinic, using the new '/ [from X.Y]' syntax.
(To be read as "positional-only from Python version X.Y")

* Issue: gh-107704

### CODE REVIEW

**Comment by erlend-aasland** [comment]
Do you want an initial review of the draft, or would you like to wait until it is marked ready for review?

**Comment by serhiy-storchaka** [comment]
I kept it in a draft state for the first run of tests and the final stage of self-review.

It is now ready for review.

File: Doc/howto/clinic.rst
Comment (erlend-aasland) [inline Doc/howto/clinic.rst]: I'm not sure this sentence is needed. It seems out of place in my opinion.

File: Lib/test/test_clinic.py
Comment (erlend-aasland) [inline Lib/test/test_clinic.py:1796]: We should also have a test for this case:

```
fn
    a: int
    / [from 3.14]
    b: int
    / [from 3.15]
```

File: Lib/test/test_clinic.py
Comment (erlend-aasland) [inline Lib/test/test_clinic.py]: I think a slightly better wording for the error message would be:

_"Function 'bar' specifies '/ [from ...]' without preceding parameters"_

File: Lib/test/test_clinic.py
Comment (erlend-aasland) [inline Lib/test/test_clinic.py]: This error message seems like a bug in the parameter parsing state machine to me; we should track that down and instead produce a message like this:

_"Function 'bar' specifies '/ [from ...]' without preceding parameters"_

File: Lib/test/test_clinic.py
Comment (erlend-aasland) [inline Lib/test/test_clinic.py]: We should improve the error message for this case as well (perhaps out of scope for this PR). IMO, this would be better:

_"Function 'bar' specifies '* [from ...]' without following parameters"_

...or:

_"Function 'bar': '* [from ...]' must come after '/'"_

File: Lib/test/test_clinic.py
Comment (erlend-aasland) [inline Lib/test/test_clinic.py:2753]: We could reuse some of the helper code here. Perhaps it is beneficial to relax the regex rules. We also need a helper to check both kinds of deprecations for a single method/function (`check_depr_kwd_and_pos`).

File: Lib/test/test_clinic.py
Comment (erlend-aasland) [inline Lib/test/test_clinic.py:3230]: This is nicer; thanks you.

File: Lib/test/test_clinic.py
Comment (erlend-aasland) [inline Lib/test/test_clinic.py:3378]: I'm not sure the `TypeError` tests are needed.

File: Modules/clinic/_testclinic_depr_star.c.h
Comment (erlend-aasland) [inline Modules/clinic/_testclinic_depr_star.c.h:1]: Let's use the opportunity to rename this to `Modules/clinic/_testclinic_depr_arg.c.h` (or a better suited name).

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]: Actually, we could drop the _"In {source}"_ part; the C preprocessor will tell us the name of the file when it generates the warning/error anyway, so it is redundant information. Only the function name should be sufficient.

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]: It is unfortunate to have to duplicate this much code between `deprecate_positional_use` and `deprecate_keyword_use`. I think it would be beneficial to refactor this further.

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]:
```suggestion
```

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]: I'm not sure this rewrite made the code easier to read; as far as I can see, it still produces the same output, so I'm not sure what's the added benefit is here.

File: Lib/test/test_clinic.py
Comment (erlend-aasland) [inline Lib/test/test_clinic.py:1]: Out of scope for this PR, but I think we should also verify that parameters are parsed correctly in the depr tests. We could use `PyTuple_Pack` instead of `Py_RETURN_NONE`, and validate that we're getting back what we're passing to the test functions.

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py:1025]: We should also produce a deprecation note for the docstring, like we do in `deprecate_positional_use()`.

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]: Regarding `_Py_ID`:

- See commit 6f6a4e6cc5cd76af4a53ffbb62b686142646ac9a
- gh-95860
- gh-90928
- gh-95907

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]:
```suggestion
            parser_definition = compiler_warning + "\n\n" + parser_definition
```

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]: IMO, the fail message is not specific enough. This would be better:

```diff
diff --git a/Tools/clinic/clinic.py b/Tools/clinic/clinic.py
index 4de45e761a..4bff0d5f18 100755
--- a/Tools/clinic/clinic.py
+++ b/Tools/clinic/clinic.py
@@ -5560,18 +5560,23 @@ def parse_slash(self, function: Function, version: VersionTuple | None) -> None:
         """Parse positional-only parameter marker '/'."""
         if version is None:
             if self.deprecated_keyword:
-                fail(f"Function {function.full_name!r}: '/ [from ...]' must come after '/'")
+                fail(f"Function {function.name!r}: '/' must precede '/ [from ...]'")
+            if self.deprecated_positional:
+                fail(f"Function {function.name!r}: '/' must precede '* [from ...]'")
+            if self.keyword_only:
+                fail(f"Function {function.name!r}: '/' must precede '*'")
             if self.positional_only:
                 fail(f"Function {function.name!r} uses '/' more than once.")
         else:
             if self.deprecated_keyword:
                 fail(f"Function {function.name!r} uses '/ [from ...]' more than once.")
+            if self.deprecated_positional:
+                fail(f"Function {function.name!r}: '/ [from ...]' must precede '* [from ...]'")
+            if self.keyword_only:
+                fail(f"Function {function.name!r}: '/ [from ...]' must precede '*'")
 
         self.positional_only = True
         self.deprecated_keyword = version
-        if self.keyword_only or self.deprecated_positional:
-            fail(f"Function {function.name!r} mixes keyword-only and "
-                 f"positional-only parameters, which is unsupported.")
         if version is not None:
             found = False
             for p in reversed(function.parameters.values()):
```

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]:
```suggestion
                fail(f"Function {function.name!r} specifies '/ [from ...]' "
                     f"without preceding parameters.")
```

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]:
```suggestion
        fail(f"Function {fname!r} specifies {symbol!r} "
                "without following parameters.", line_number=lineno)
```

or:

```suggestion
        fail(f"Function {fname!r} specifies {symbol!r} "
                "without succeeding parameters.", line_number=lineno)
```

**Review by erlend-aasland — COMMENTED** [review]
I think it would be beneficial to try and reduce some of the code duplication, either in this PR, or in a following-up refactor. Perhaps the latter is preferable.

File: Doc/howto/clinic.rst
Comment (serhiy-storchaka) [reply Doc/howto/clinic.rst]: Removed.

File: Lib/test/test_clinic.py
Comment (serhiy-storchaka) [reply Lib/test/test_clinic.py:1796]: Done. Also for `* [from ...]`.

File: Lib/test/test_clinic.py
Comment (serhiy-storchaka) [reply Lib/test/test_clinic.py]: Done.

File: Lib/test/test_clinic.py
Comment (serhiy-storchaka) [reply Lib/test/test_clinic.py]: Yes, this error is actually raised for `/` (see also `test_single_slash`) and it is considered as a bug. I'll remove this test.

File: Lib/test/test_clinic.py
Comment (serhiy-storchaka) [reply Lib/test/test_clinic.py]: But it will lead to repeating the check and the error message 4 times with tiny modifications for all combinations of `*` and `/`. Maybe more if support multiple versions.

File: Lib/test/test_clinic.py
Comment (serhiy-storchaka) [reply Lib/test/test_clinic.py:3378]: It is here to test that there is no deprecation warning if this kind or TypeError is raised. My earliest version emitted it, and one of your intermediate versions emitted a warning in a similar case (before adding the check for the upper limit).

File: Modules/clinic/_testclinic_depr_star.c.h
Comment (serhiy-storchaka) [reply Modules/clinic/_testclinic_depr_star.c.h:1]: Will do it later. I'm afraid that it might make the review process more difficult.

File: Tools/clinic/clinic.py
Comment (serhiy-storchaka) [reply Tools/clinic/clinic.py]: Done. Then we can drop also the call of `wrapped_c_string_literal()`.

File: Tools/clinic/clinic.py
Comment (serhiy-storchaka) [reply Tools/clinic/clinic.py]: It slightly reduced the number of lines 21->18, and I hoped to make it more similar to the message in `deprecate_keyword_use()`. But there are too many differences anyway.

Restored the old code.

File: Tools/clinic/clinic.py
Comment (serhiy-storchaka) [reply Tools/clinic/clinic.py]: Done.

File: Tools/clinic/clinic.py
Comment (serhiy-storchaka) [reply Tools/clinic/clinic.py]: Unfortunately it is the simplest general way to check that the argument was passed by keyword. It is not the fastest. For inlined parsing code we can use more efficient but more complex code, we can even build it in `_PyArg_UnpackKeywords()`. But if `PyArg_ParseTupleAndKeywords()` or similar function is used, the only other way is to use deprecated and inefficient `PyMapping_HasKeyString()` or introduce a new private inefficient API.

File: Tools/clinic/clinic.py
Comment (serhiy-storchaka) [reply Tools/clinic/clinic.py:1025]: Ah, I missed it. It was added after I cloned `deprecate_positional_use()`.

File: Tools/clinic/clinic.py
Comment (serhiy-storchaka) [reply Tools/clinic/clinic.py]: Done.

File: Tools/clinic/clinic.py
Comment (serhiy-storchaka) [reply Tools/clinic/clinic.py]: Done.

File: Tools/clinic/clinic.py
Comment (serhiy-storchaka) [reply Tools/clinic/clinic.py]: Done.

File: Tools/clinic/clinic.py
Comment (serhiy-storchaka) [reply Tools/clinic/clinic.py]: Done.

File: Lib/test/test_clinic.py
Comment (serhiy-storchaka) [reply Lib/test/test_clinic.py:2753]: If so much effort was spent for formatting warning message ("a parameter" or "parameters" instead of simple "parameter(s)", etc), it is a waste that the check is so lenient. It can miss some actual errors.

File: Tools/clinic/clinic.py
Comment (serhiy-storchaka) [reply Tools/clinic/clinic.py]: After adding support of different versions all this code will be rewritten. If there still be some similarity, it will be very different.

**Review by serhiy-storchaka — COMMENTED** [review]
Thank you for your review @erlend-aasland. I tried to address them. Seems the only unresolved issue is with `_Py_ID()`.

File: Modules/clinic/_testclinic_depr_star.c.h
Comment (erlend-aasland) [reply Modules/clinic/_testclinic_depr_star.c.h:1]: Sure, we can do it in a follow-up PR.

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [inline Tools/clinic/clinic.py]: We should decide which variant to chose: `function.full_name` or `function.name`. IMO, `.name` should be sufficient. After I improved the accuracy of the line numbers, it now very easy to locate the offending line.

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: Thanks, that made the diff smaller, and the review easier.

File: Tools/clinic/clinic.py
Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: cc. @ericsnowcurrently

Comment (ericsnowcurrently) [reply Tools/clinic/clinic.py]: What's the concern with using `_Py_ID()` here?

Comment (erlend-aasland) [reply Tools/clinic/clinic.py]: Currently, there is no concern; I just wanted to point out that even though Argument Clinic does not currently care about the Limited API, we _may_ want to add such functionality in the future; IMO, it should be possible for Argument Clinic to generate code that does not depend on internal APIs, for example for stdlib extension modules such as sqlite3.

Many core devs seem to be in favour of not using internal APIs in stdlib extension modules, but I'm not sure if there is a consensus.

**Review by erlend-aasland — APPROVED** [review]
Thanks for doing this, @serhiy-storchaka. @AlexWaygood, do you want to have a look?

File: Doc/howto/clinic.rst
Comment (AlexWaygood) [inline Doc/howto/clinic.rst]:
```suggestion
however, we'll have to wait two releases before making these changes,
```

File: Doc/howto/clinic.rst
Comment (AlexWaygood) [inline Doc/howto/clinic.rst]:
```suggestion
them positional-only and keyword-only respectively in Python 3.14 at
```

File: Doc/howto/clinic.rst
Comment (AlexWaygood) [inline Doc/howto/clinic.rst]:
```suggestion
below the *b* parameter and adding the line ``* [from 3.14]`` right above
```

**Review by AlexWaygood — COMMENTED** [review]
Some small suggestions for the docs:

File: Tools/clinic/clinic.py
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:5505]: Could you add a one-line docstring here explaining what this function does, similar to the other `parse_foo` methods on this class?

File: Tools/clinic/clinic.py
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]: Could you add a note to this docstring explaining what the `version` parameter signifies?

File: Tools/clinic/clinic.py
Comment (AlexWaygood) [inline Tools/clinic/clinic.py]: Same here, it would be nice to explain the significance of the `version` parameter for new readers of the code

**Review by AlexWaygood — COMMENTED** [review]
Code looks good, just a few very minor nits:

File: Tools/clinic/clinic.py
Comment (serhiy-storchaka) [inline Tools/clinic/clinic.py:5505]:
```suggestion
    def parse_version(self, thenceforth: str) -> VersionTuple:
        """Parse Python version in `[from ...]` marker."""
```

File: Tools/clinic/clinic.py
Comment (serhiy-storchaka) [inline Tools/clinic/clinic.py]:
```suggestion
        """Parse keyword-only parameter marker '*'.

        The 'version' parameter signifies the future version from which
        the marker will take effect (None means it is already in effect).
        """
```

File: Tools/clinic/clinic.py
Comment (serhiy-storchaka) [inline Tools/clinic/clinic.py]:
```suggestion
        """Parse positional-only parameter marker '/'.

        The 'version' parameter signifies the future version from which
        the marker will take effect (None means it is already in effect).
        """
```

**Review by serhiy-storchaka — COMMENTED** [review]
Thank you for your review @AlexWaygood. Do these docstrings work?

**Review by AlexWaygood — APPROVED** [review]
LGTM. Thanks @serhiy-storchaka!

**Comment by serhiy-storchaka** [comment]
Post-review changes:

* I have found how to make keyword argument check more efficient in case of inlined parsing code. It now has almost zero overhead.
* Since the tests with the new generated code no longer covered all cases in the generator, I added new tests, and they exposed bugs in a corner case in constructors (including `* [from ...]`). Fixed.

**Pull request merged**
